New "Filled line" symbol layer type

nyalldawson opened the pull request:
This adds a new line symbol type which renders lines using a fill symbol. The interior of the line is drawn using any standard QGIS fill symbol, allowing for lines filled with gradients, line hatches, etc.

Sponsored by North Road, thanks to SLYR

![Peek 2023-11-04 14-37](https://github.com/qgis/QGIS/assets/1829991/0fbf6560-5d44-4ec4-b250-bccab2866969)

nirvn commented on src/core/symbology/qgslinesymbollayer.cpp:
```diff
@@ -3788,3 +3789,309 @@ void QgsLineburstSymbolLayer::setColorRamp( QgsColorRamp *ramp )
+  {
+    width = props[QStringLiteral( "outline_width" )].toDouble();
+  }
+  else if ( props.contains( QStringLiteral( "width" ) ) )
+  {
+    //pre 2.5 projects used "width"
+    width = props[QStringLiteral( "width" )].toDouble();
+  }
```
Do we need this for a new symbol layer that? We'll never see a pre 2.5 project with this symbol layer type ever :wink: . If you were thinking of adding it as an alias, maybe change the comment?

nirvn commented on src/core/symbology/qgslinesymbollayer.cpp:
```diff
@@ -3788,3 +3789,309 @@ void QgsLineburstSymbolLayer::setColorRamp( QgsColorRamp *ramp )
+  {
+    l->setWidthUnit( QgsUnitTypes::decodeRenderUnit( props[QStringLiteral( "outline_width_unit" )].toString() ) );
+  }
+  else if ( props.contains( QStringLiteral( "width_unit" ) ) )
+  {
+    //pre 2.5 projects used "width_unit"
+    l->setWidthUnit( QgsUnitTypes::decodeRenderUnit( props[QStringLiteral( "width_unit" )].toString() ) );
+  }
```
Same Q.

nirvn commented on tests/src/python/test_qgsfilledlinesymbollayer.py:
```diff
@@ -0,0 +1,160 @@
+"""
+
+__author__ = 'Nyall Dawson'
+__date__ = 'October 2021'
+__copyright__ = '(C) 2021, Nyall Dawson'
```
```suggestion
__date__ = 'November 2023'
__copyright__ = '(C) 2023, Nyall Dawson'
```

nirvn reviewed (approved):
Nice one.

nyalldawson committed:
New "Filled line" symbol layer type

This adds a new line symbol type which renders lines using a
fill symbol. The interior of the line is drawn using any standard
QGIS fill symbol, allowing for lines filled with gradients, line
hatches, etc.

Sponsored by North Road, thanks to SLYR

nyalldawson committed:
Update comment

nyalldawson committed:
Update date
